server/src/main/java/org/candlepin/resource/OwnerContentResource.java

```diff
@@ -276,10 +289,13 @@ public ContentDTO updateContent(@PathParam("owner_key") String ownerKey,
             throw new ForbiddenException(i18n.tr("content \"{0}\" is locked", existing.getId()));
         }
 
-        existing = this.contentManager.updateContent(owner, content, true);
+        Content updated = new EntityVersioningRetryWrapper()
+            .attempts(VERSIONING_CONSTRAINT_VIOLATION_ATTEMPTS)
+            .execute(() -> this.contentManager.updateContent(owner, content, true));
```

Comment:
If the retry logic is going to need to be wrapped around all of our Product/Content CRUD operations (or at least most of them), then I wouldn't let that be the burden of the caller. There must be a way to enforce this as the default behavior of the ContentManager/ProductManager methods, and then give the caller a way to opt-out of it for the few times we don't need to retry?

Reply:
Yeah... this is a tough one. I generally agree that this kind of thing should be completely transparent to the callers, but where things started going off the rails here is in that the ContentManager calls into the ProductManager to update affected products when content changes. We wouldn't want the retry to be multi-layered, and I wasn't sure if I wanted to pollute the namespace with a bunch of methods that would only be called by the ContentManager.

The original prototype had this concept that extended the two managers and had methods that did nothing but wrap the call to the superclass's version in the retry logic. It worked, but it was a bit jank looking and feeling, and hit the issue of nesting those transactional objects during content updates, so I dropped it and went with what we have here. The managers being singletons might get in the way of this, but maybe something along the lines of having the standard binding to the retry-wrapped subclass, and the ContentManager uses a binding to get the vanilla ProductManager without retry logic?

Definitely open to suggestions and thoughts here.

Reply:
I thought Guice by default uses new instances if you don't specify scope, and I don't see a Singleton annotation or binding call in CandlepinModule.java for the manager classes, so I don't think they're singletons. Even if they were, Guice singletons aren't true singletons, but more like limited-instantiation objects, so you could just have 2 instances only of them if you use the Named annotation to differentiate between them.

Also, you could entirely avoid having 2 different manager classes for retry/non-retry, by embbedding the retry logic in the single manager class, and have it be configurable (so that you can set the number of retries to 0), and then defining 2 guice provider methods with Provides and Named annotations, where one returns the manager class configured with retries, and the other without retries (we already do this kind of thing for ObjectMapper classes if you check CandlepinModule.java)

Reply:
Did we remove the `@Singleton` annotation from the managers? I could have sworn they were flagged as such at one point (and, honestly, they probably should be).

I'll poke around a bit with some implementations and post a gist or two or a commit here and we'll see how they feel.

Reply:
These have all been dropped for the moment, but I'm still conflicted. Given some of the other transaction problems, this might end up getting resolved once I decide on what to do there.

Reply:
Side note on the singleton thing: there are some context issues that arise when flagging either manager as a singleton. Probably worth investigating eventually, but not part of this task.